webpack/webpack-embedded-dependencies-plugin/src/index.ts

```diff
@@ -0,0 +1,282 @@
+  author?: string | { name?: string };
+  licenses?: { type: string; url: string }[];
+  licensePath?: string;
+}
+
+interface IEmbeddedDependenciesWebpackPluginOptions {
+  outputFileName?: string;
+  generateLicenseFile?: boolean;
+  generateLicenseFileFunction?: LicenseFileGeneratorFunction;
+  generatedLicenseFilename?: LicenseFileName;
+}
+
+type LicenseFileGeneratorFunction = (packages: IPackageData[]) => string;
+type PackageMapKey = `${string}@${string}`;
+type LicenseFileName = `${string}.${'html' | 'md' | 'txt'}`;
+type ThirdPartyPackageMap = Map<PackageMapKey, { dir: string; data: IPackageData }>;
+type FlattenedPackageEntry = [PackageMapKey, { dir: string; data: IPackageData }];
+type LicenseSourceCache = Map<string, string>;
+
+/**
+ * @alpha
+ * Webpack plugin that generates a file with the list of embedded dependencies
+ * and their licenses.
+ */
+export default class EmbeddedDependenciesWebpackPlugin implements WebpackPluginInstance {
+  public outputFileName: string;
+  public generateLicenseFile: boolean;
+  public generateLicenseFileFunction: LicenseFileGeneratorFunction;
+  public generatedLicenseFilename: LicenseFileName;
+  private _licenseSourceCache: LicenseSourceCache = new Map();
+
+  public constructor(options?: IEmbeddedDependenciesWebpackPluginOptions) {
+    this.outputFileName = options?.outputFileName || 'embedded-dependencies.json';
+    this.generateLicenseFile = options?.generateLicenseFile || false;
+    this.generateLicenseFileFunction =
+      options?.generateLicenseFileFunction || this._defaultLicenseFileGenerator;
+    this.generatedLicenseFilename = options?.generatedLicenseFilename || 'THIRD-PARTY-NOTICES.html';
+  }
+
+  public apply(compiler: Compiler): void {
+    const thirdPartyPackages: ThirdPartyPackageMap = new Map();
+
+    compiler.hooks.normalModuleFactory.tap(PLUGIN_NAME, (normalModuleFactory) => {
+      normalModuleFactory.hooks.module.tap(
+        PLUGIN_NAME,
+        (module, moduleCreateData: IWebpackModuleCreateData, resolveData) => {
+          const { resourceResolveData } = moduleCreateData;
+          const pkg: IPackageData | undefined = resourceResolveData?.descriptionFileData;
+          const filePath: string | undefined = resourceResolveData?.descriptionFileRoot;
+
+          if (pkg && filePath && filePath.includes('node_modules')) {
+            const key: PackageMapKey = makePackageMapKeyForPackage(pkg);
+            thirdPartyPackages.set(key, { dir: filePath, data: pkg });
+          }
+
+          return module;
+        }
+      );
+    });
+
+    // Tap into compilation so we can tap into compilation.hooks.processAssets
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.processAssets.tapPromise(
+        { name: PLUGIN_NAME, stage: Compilation.PROCESS_ASSETS_STAGE_REPORT },
+        async (assets) => {
+          const rawPackages: FlattenedPackageEntry[] = Array.from(thirdPartyPackages).sort((first, next) => {
```

Comment:
Per discussion, makes more sense to postpone the sorting until after the async processing happens.